### PR TITLE
fix: mitigate DID pallet replay vulnerabilities

### PR DESCRIPTION
**Pull request by Diiaablo95**
## fixes https://github.com/KILTprotocol/srlabs-audit/issues/8

This PR introduces the following features/fixes:

- **Transaction mortality**: All DID-authorised operations (hence excluding DID creation) can only be submitted to the blockchain after a maximum of `MaxBlocksTxValidity` after their creation (currently set to 1h time). This means that when a DID operation is evaluated, the blockchain current block number must fall in the inclusive range `[operation_block, operation_block + MaxBlocksTxValidity]`. Anything falling outside of this range results in an error.
- **Nonces wrapping**: Before, when a nonce (hypothetically) reached its end of life, meaning its maximum value, a DID became (still theoretically) unusable because wrapping the nonce would introduce replay attacks vulnerabilities. Now that we have transaction mortality nonces can safely wrap around with no (hopefully) security downside.
- **One-time DID creation**: Once a DID is deleted, we assume that all the keys that have ever been under the control of the DID subject have been compromised, including the initial account key that was used to write the DID on chain. If we would not prevent a DID from being recreated, an attacker owning a compromised account key might just wait for the DID to be deleted before generating a new valid signed creation operation and technically taking control of someone else's DID. Even though this is an edge case, it still provides additional security guarantees that once deleted, nobody can ever use that DID again.

## Checklist

- [x] Implement features
- [x] Add/update unit tests
- [x] Add/update benchmarks
- [x] Add/update comments
- [x] Get it reviewed internally
- [x] Get it reviewed by SRLabs -> https://github.com/KILTprotocol/srlabs-audit/issues/8#issuecomment-931275237
- [x] Re-execute benchmarks
- [x] Bump runtime version -> runtime # 26
- [x] Open companion types PR -> https://github.com/KILTprotocol/type-definitions/pull/24
- [x] Update SDK to work with latest types - piggybacking onto https://github.com/KILTprotocol/sdk-js/pull/415 -> Commit: https://github.com/KILTprotocol/sdk-js/pull/415/commits/e462b2166d2871fd7a4c99b32f887a81bed31de2

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] This PR does not introduce new custom types
  - [x] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

**Comment by Diiaablo95** [comment]
@weichweich @wischli I would like the PR to be reviewed internally, then to pass it through SRLabs, then re-run the benchmarks and all the other routine tasks before merging into `develop`.

File: pallets/did/src/tests.rs [hunk 1]
```diff
@@ -1670,6 +1679,81 @@ fn check_too_large_tx_counter_call_error() {
...
+
+	// But it would work if the system would be one block earlier.
+	ext.execute_with(|| {
+		System::set_block_number(call_operation.operation.block_number + MaxBlocksTxValidity::get());
+		assert_ok!(Did::submit_did_call(
+			Origin::signed(caller),
+			Box::new(call_operation.operation),
+			did::DidSignature::from(signature)
+		),);
+	});
```
Comment (wischli) [inline pallets/did/src/tests.rs:1720]: W: See above regarding multiple Externalities.

```suggestion
	ExtBuilder::default()
		.with_dids(vec![(did.clone(), mock_did)])
		.build(None).
		execute_with(|| {
			let submitter = kilt_primitives::AccountId::default();
			let call_operation = generate_test_did_call(did::DidVerificationKeyRelationship::Authentication, did, submitter);
			let signature = auth_key.sign(call_operation.encode().as_ref());
					
			// System block number 1 past the max block the operation was allowed for.
			System::set_block_number(call_operation.operation.block_number + MaxBlocksTxValidity::get() + 1);
			assert_noop!(
				Did::submit_did_call(
					Origin::signed(caller.clone()),
					Box::new(call_operation.operation.clone()),
					did::DidSignature::from(signature.clone())
				),
				did::Error::<Test>::InvalidOperationValidity
			);

			// But it would work if the system would be one block earlier.
			System::set_block_number(call_operation.operation.block_number + MaxBlocksTxValidity::get());
			assert_ok!(Did::submit_did_call(
				Origin::signed(caller),
				Box::new(call_operation.operation),
				did::DidSignature::from(signature)
			),);
		});
```

File: pallets/did/src/tests.rs [hunk 2]
```diff
@@ -1670,6 +1679,81 @@ fn check_too_large_tx_counter_call_error() {
...
+		System::set_block_number(call_operation.operation.block_number - MaxBlocksTxValidity::get() - 1);
+		assert_noop!(
+			Did::submit_did_call(
+				Origin::signed(caller.clone()),
+				Box::new(call_operation.operation.clone()),
+				did::DidSignature::from(signature.clone())
+			),
+			did::Error::<Test>::InvalidOperationValidity
+		);
+	});
```
Comment (wischli) [inline pallets/did/src/tests.rs:1754]: W: Wrap all of this into the execution of the Externality.

File: runtimes/peregrine/src/lib.rs [hunk 3]
```diff
@@ -590,6 +590,8 @@ parameter_types! {
 	pub const MaxTotalKeyAgreementKeys: u32 = 1000;
 	#[derive(Debug, Clone, PartialEq)]
 	pub const MaxEndpointUrlsCount: u32 = 3u32;
+	// 300 blocks * 12s/block = 3.600s = 1h
+	pub const MaxBlocksTxValidity: u64 = 300u64;
```
Comment (wischli) [inline runtimes/peregrine/src/lib.rs:594]: S: Add to primitives and use type `BlockNumber` instead of `u64` as you defined `type MaxBlocksTxValidity: Get<BlockNumberOf<Self>>;`.

File: runtimes/standalone/src/lib.rs [hunk 4]
```diff
@@ -389,6 +389,8 @@ parameter_types! {
 	pub const MaxTotalKeyAgreementKeys: u32 = 1000;
 	#[derive(Debug, Clone, PartialEq)]
 	pub const MaxEndpointUrlsCount: u32 = 3u32;
+	// 600 blocks * 6s/block = 3.600s = 1h
+	pub const MaxBlocksTxValidity: u64 = 600u64;
```
Comment (wischli) [inline runtimes/standalone/src/lib.rs:393]: W: I see why you might have not added it to the primitives as I requested above. I would still do that for `peregrine` and soon `spiritnet` as well as keep this definition here separate for `standalone`. 

In any case, please use `BlockNumber` type.

**Review by wischli — CHANGES_REQUESTED** [review]
Overall, the PR looks good. I have a few comments regarding formatting, your test-style and the newly added pallet parameter.

File: pallets/did/src/lib.rs [hunk 5]
```diff
@@ -256,6 +279,15 @@ pub mod pallet {
...
 	pub type Did<T> = StorageMap<_, Blake2_128Concat, DidIdentifierOf<T>, DidDetails<T>>;
 
+	/// The set of DIDs that have been deleted and cannot therefore be created
+	/// again for security reasons.
+	///
+	/// It maps from a DID identifier to a unit tuple, for the sake of tracking
+	/// DID identifiers.
+	#[pallet::storage]
+	#[pallet::getter(fn get_deleted_did)]
+	pub(crate) type DeletedDids<T> = StorageMap<_, Blake2_128Concat, DidIdentifierOf<T>, ()>;
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/lib.rs:289]: Yeah this storage is for the system to work correctly, so it should not concern users. I am not 100% in the deposit discussion, but I think this would be stored on the chain indefinitely without a corresponding deposit, as also Shawn suggested. The deposit should be given fully back to the user when they delete their DID.

File: pallets/did/src/lib.rs [hunk 6]
```diff
@@ -902,13 +954,24 @@ impl<T: Config> Pallet<T> {
...
+		let current_block_number = <frame_system::Pallet<T>>::block_number();
+		// Before accessing the storage, we check if the provided block number is valid,
+		// i.e., if the current blockchain block is in the inclusive range
+		// [operation_block_number, operation_block_number + MaxBlocksTxValidity].
+		let allowed_range =
+			operation.block_number..=operation.block_number.saturating_add(T::MaxBlocksTxValidity::get());
+		ensure!(
+			allowed_range.contains(&current_block_number),
+			DidError::SignatureError(SignatureError::InvalidBlockNumber)
+		);
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/lib.rs:966]: I think a range is easier to understand, honestly, so I would rather keep it this way, especially since you don't seem to have a strong opinion about it either.

File: pallets/did/src/tests.rs [hunk 7]
```diff
@@ -228,6 +228,24 @@ fn check_duplicate_did_creation() {
...
+	let signature = auth_key.sign(details.encode().as_ref());
+
+	let mut ext = ExtBuilder::default().with_deleted_dids(vec![alice_did]).build(None);
+
+	ext.execute_with(|| {
+		assert_noop!(
+			Did::create(Origin::signed(ACCOUNT_00), details, did::DidSignature::from(signature),),
+			did::Error::<Test>::DidAlreadyDeleted
+		);
+	});
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/tests.rs:246]: But `ext` is always the same test externality for each use case. I would like to know why this approach is not ideal and especially why the suggested one is better :) My point is that whatever does not need to be executed within an externality is not executed in an externality, but I am willing to know your opinion about it.

File: pallets/did/src/tests.rs (hunk 1 above)
Comment (Diiaablo95) [reply pallets/did/src/tests.rs:1720]: I guess as you said it will be the scope of a different PR in the future

File: pallets/did/src/tests.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/tests.rs:246]: > whatever does not need to be executed within an externality is not executed in an externality

wrapping everything in `ext.execute_with(|| {})` comes with a huge readability cost. My question would be why we would want to move everything out of the execute with block. What is the benefit?

[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/tests.rs:246]: It makes clear, for me at least, what is part of the initialisation/verification process, which does not require a context to run in, and what is actually being executed in the test environment. But if you guys have already agreed the other way is the way to go, I guess it is okay then and I will remember that next time.

File: runtimes/standalone/src/lib.rs (hunk 4 above)
Comment (Diiaablo95) [reply runtimes/standalone/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/c471802786366c73e670f33afc75d8498d1a0758

File: runtimes/peregrine/src/lib.rs (hunk 3 above)
Comment (Diiaablo95) [reply runtimes/peregrine/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/c471802786366c73e670f33afc75d8498d1a0758

File: pallets/did/src/lib.rs [hunk 8]
```diff
@@ -787,6 +835,9 @@ pub mod pallet {
 				<Error<T>>::DidNotPresent
 			);
 
+			// Marking them as deleted and they're gone forever 👋👋👋
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/c471802786366c73e670f33afc75d8498d1a0758

File: pallets/did/src/lib.rs [hunk 9]
```diff
@@ -94,6 +107,11 @@
 //! - The chain performs basic checks over the endpoint URLs provided in
 //!   creation and deletion operations. The SDK will perform more in-depth
 //!   validation of the URL string, e.g., by pattern-matching using regexes.
+//!   - After it is generated and signed by a client, a DID-authorised operation
+//!     can
+//!   be submitted for evaluation anytime between the time the operation is
+//! created   and [MaxBlocksTxValidity] blocks after that. After this time has
+//! elapsed,   the operation is considered invalid.
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/c471802786366c73e670f33afc75d8498d1a0758

File: pallets/did/src/tests.rs (hunk 1 above)
Comment (wischli) [reply pallets/did/src/tests.rs:1720]: Yes.

File: runtimes/peregrine/src/lib.rs (hunk 3 above)
Comment (wischli) [reply runtimes/peregrine/src/lib.rs]: Great improvement!

File: pallets/did/src/tests.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/tests.rs:246]: For me it makes it just harder to read.

File: pallets/did/src/tests.rs (hunk 1 above)
Comment (weichweich) [reply pallets/did/src/tests.rs:1720]: Since you touched this test, i would already apply Williams suggestion.

File: pallets/did/src/tests.rs (hunk 2 above)
Comment (Diiaablo95) [reply pallets/did/src/tests.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/639389dc8c1f3317b4a523fa6225412a21d53b01

File: pallets/did/src/tests.rs (hunk 1 above)
Comment (Diiaablo95) [reply pallets/did/src/tests.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/639389dc8c1f3317b4a523fa6225412a21d53b01

File: pallets/did/src/tests.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/tests.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/639389dc8c1f3317b4a523fa6225412a21d53b01

File: pallets/did/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/did/src/lib.rs:966]: To me, my proposal is easier to understand. Range definitely adds more complexity than just comparing two numbers.

Also, it just seems overkill to create a range for such a simple check and I would think it's a bit more resource heavy (but probably negligible small in Rust)?! 

Anyway, I do not have a strong opinion.

**Review by wischli — APPROVED** [review]
LGTM

File: pallets/did/src/tests.rs [hunk 10]
```diff
@@ -2352,17 +2384,18 @@ fn check_invalid_signature_operation_verification() {
...
-				&call_operation,
-				&did::DidSignature::from(signature)
-			),
-			did::DidError::SignatureError(did::SignatureError::InvalidSignature)
-		);
-	});
+	ExtBuilder::default()
+		.with_dids(vec![(did, mock_did)])
+		.build(None)
+		.execute_with(|| {
```
Comment (weichweich) [inline pallets/did/src/tests.rs:2390]: I really like this short version. I guess you updated all of them now? 😅 I would have only done the once you really had to change anyways but fine. 👍🏻

File: pallets/did/src/lib.rs [hunk 11]
```diff
@@ -787,6 +834,10 @@ pub mod pallet {
 				<Error<T>>::DidNotPresent
 			);
 
+			// Mark as deleted to prevent potential replay-attacks of re-adding a previously
+			// deleted DID.
+			<DeletedDids<T>>::insert(&did_subject, ());
```
Comment (weichweich) [inline pallets/did/src/lib.rs:839]:
```suggestion
			DeletedDids::<T>::insert(&did_subject, ());
```

File: pallets/did/src/lib.rs [hunk 12]
```diff
@@ -902,13 +954,24 @@ impl<T: Config> Pallet<T> {
 		operation: &DidAuthorizedCallOperationWithVerificationRelationship<T>,
 		signature: &DidSignature,
 	) -> Result<(), DidError> {
+		let current_block_number = <frame_system::Pallet<T>>::block_number();
```
Comment (weichweich) [inline pallets/did/src/lib.rs:957]:
```suggestion
		let current_block_number = frame_system::Pallet::<T>::block_number();
```

**Review by weichweich — CHANGES_REQUESTED** [review]
Looks code. Only got some style and naming requests...

File: pallets/did/src/migrations.rs [hunk 13]
```diff
@@ -48,7 +49,7 @@ impl DidStorageVersion {
 // old version anymore.
 impl Default for DidStorageVersion {
 	fn default() -> Self {
-		Self::V1
+		Self::V2
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/migrations.rs:52]: Sorry I do not understand what you mean.

File: pallets/did/src/lib.rs (hunk 12 above)
Comment (Diiaablo95) [reply pallets/did/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/e221572120083c2ad3606a1a3847d56ba0e4968b

File: pallets/did/src/lib.rs (hunk 11 above)
Comment (Diiaablo95) [reply pallets/did/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/e221572120083c2ad3606a1a3847d56ba0e4968b

File: pallets/did/src/mock.rs [hunk 14]
```diff
@@ -381,6 +381,14 @@ impl ExtBuilder {
 			});
 		}
 
+		if !self.deleted_dids.is_empty() {
+			ext.execute_with(|| {
+				self.deleted_dids.iter().for_each(|did| {
+					did::DeletedDids::<Test>::insert(did.clone(), ());
+				})
+			})
+		}
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/mock.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/e221572120083c2ad3606a1a3847d56ba0e4968b

File: pallets/did/src/errors.rs [hunk 15]
```diff
@@ -87,6 +86,8 @@ pub enum SignatureError {
 	InvalidSignature,
 	/// The operation nonce is not equal to the current DID nonce + 1.
 	InvalidNonce,
+	/// The provided operation block number is not valid.
+	InvalidBlockNumber,
```
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/errors.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/e221572120083c2ad3606a1a3847d56ba0e4968b

File: pallets/did/src/migrations.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/migrations.rs:52]: The `StorageVersion` storage is not initialized on genesis. If you change the default value for the version you will change what is read from the storage. 

* The storage is empty -> default value will be returned
* default version is current version -> no migrations are executed
* no migrations -> storage is wrong
* wrong storage -> disaster 💥

[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/did/src/migrations.rs:52]: just checked on peregrine & peregrine-stg, the version is set. So all good

**Review by weichweich — APPROVED** [review]
🏅Good!

File: pallets/did/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Diiaablo95) [reply pallets/did/src/lib.rs]: https://github.com/KILTprotocol/mashnet-node/pull/257/commits/f55b3d3c499288c02601439d6fddccb77e238104
